test/modules/core/controllers/test-controller.js

```diff
@@ -0,0 +1,195 @@
+  },
+  {
+    title: 'wheel#out of bounds',
+    props: {x: 200},
+    events: makeEvents(['wheel']),
+    viewStateChanges: 0,
+    interactionStateChanges: 0
+  },
+  {
+    title: 'wheel#disabled',
+    props: {scrollZoom: false},
+    events: makeEvents(['wheel']),
+    viewStateChanges: 0,
+    interactionStateChanges: 0
+  },
+
+  {
+    title: 'doubletap',
+    props: {},
+    events: makeEvents(['doubletap']),
+    viewStateChanges: 1,
+    interactionStateChanges: 1
+  },
+  {
+    title: 'doubletap#out of bounds',
+    props: {x: 200},
+    events: makeEvents(['doubletap']),
+    viewStateChanges: 0,
+    interactionStateChanges: 0
+  },
+  {
+    title: 'doubletap#disabled',
+    props: {doubleClickZoom: false},
+    events: makeEvents(['doubletap']),
+    viewStateChanges: 0,
+    interactionStateChanges: 0
+  },
+
+  {
+    title: 'keyboard',
+    props: {},
+    events: makeEvents(['keydown'], {keyCode: 189})
+      .concat(makeEvents(['keydown'], {keyCode: 189, shiftKey: true}))
+      .concat(makeEvents(['keydown'], {keyCode: 187}))
+      .concat(makeEvents(['keydown'], {keyCode: 187, shiftKey: true}))
+      .concat(makeEvents(['keydown'], {keyCode: 37}))
+      .concat(makeEvents(['keydown'], {keyCode: 37, shiftKey: true}))
+      .concat(makeEvents(['keydown'], {keyCode: 38}))
+      .concat(makeEvents(['keydown'], {keyCode: 38, shiftKey: true}))
+      .concat(makeEvents(['keydown'], {keyCode: 39}))
+      .concat(makeEvents(['keydown'], {keyCode: 39, shiftKey: true}))
+      .concat(makeEvents(['keydown'], {keyCode: 40}))
+      .concat(makeEvents(['keydown'], {keyCode: 40, shiftKey: true})),
+    viewStateChanges: 12,
+    interactionStateChanges: 12
+  },
+  {
+    title: 'keyboard#disabled',
+    props: {keyboard: false},
+    events: makeEvents(['keydown'], {keyCode: 189}),
+    viewStateChanges: 0,
+    interactionStateChanges: 0
+  }
+];
+
+export default function testController(t, ViewClass, defaultProps, blackList = []) {
+  let onViewStateChangeCalled = 0;
+  let onStateChangeCalled = 0;
+  const onViewStateChange = () => onViewStateChangeCalled++;
+  const onStateChange = () => onStateChangeCalled++;
+
+  const controllerProps = new ViewClass({controller: true}).controller;
+  Object.assign(
+    defaultProps,
+    BASE_PROPS,
+    {
+      onViewStateChange,
+      onStateChange
+    },
+    controllerProps
+  );
+  const ControllerClass = controllerProps.type;
+  const controller = new ControllerClass(defaultProps);
+
+  for (const testCase of TEST_CASES) {
+    if (blackList.includes(testCase.title)) {
+      continue; // eslint-disable-line
+    }
+    // reset
+    onViewStateChangeCalled = 0;
+    onStateChangeCalled = 0;
+
+    controller.setProps(Object.assign({}, defaultProps, testCase.props));
+    for (const event of testCase.events) {
+      controller.handleEvent(event);
+    }
+    t.is(onViewStateChangeCalled, testCase.viewStateChanges, `${testCase.title} onViewStateChange`);
```

Comment:
This is good enough, but I wonder if it is possible to check the specific events in sequence which are supposed to be triggered in each case.